### PR TITLE
Add Login acceptance test

### PR DESCRIPTION
**Pull request by bibekthapa619**
# Pull Request Template

## Description
Different scenarios of user login has been added.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested

- [ ] Unit Test
- [ ] Integration Test

## Checklist

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: frontend/tests/acceptance/features/webUILogin/login.feature [hunk 1]
```diff
@@ -1,9 +1,35 @@
...
+          And the user has browsed to the login page
+
+     @validLogin
+     Scenario: login a valid user
+          When the user logs in with email "user7771@gmail.com" and password "password" using the webUI
+          Then the user should be redirected to the dashboard page
+
+     @invalidLogin
+     Scenario: login with invalid credentials
+          When user logs in with following credentials:
```
Comment (phil-davis) [inline frontend/tests/acceptance/features/webUILogin/login.feature:17]:
```suggestion
          When the user logs in with the following credentials:
```

File: frontend/tests/acceptance/features/webUILogin/login.feature [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
+               | email    | user1@gmail.com |
```
Comment (phil-davis) [inline frontend/tests/acceptance/features/webUILogin/login.feature:18]:
```suggestion
               | email    | user7771@gmail.com |
```

I guess this scenario is to test login with a valid email but invalid password?

Or you could make 4 scenarios:

Scenario: login with invalid credentials (unknown email and invalid password) - this scenario
Scenario: login with unknown email but valid password of some other user
Scenario: login with valid email but invalid password
Scenario: login with valid email and valid password of some other user (needs 2 users singed-up, and check that you cannot login using your own email and the other user's password)

File: frontend/tests/acceptance/step_definitions/loginContext.js [hunk 3]
```diff
@@ -1,19 +1,54 @@
...
+Then("the user should be redirected to the dashboard page", () =>
+  I.seeElement(dashboard.dashboardContainer)
+);
 
-When('the user logs in with email {string} and password {string} using the webUI', (email, password) => {
+Given("the user has browsed to the login page", () => {
+  I.amOnPage(loginPage.url);
+});
+
+When("user logs in with following credentials:", (table) => {
```
Comment (phil-davis) [inline frontend/tests/acceptance/step_definitions/loginContext.js:34]:
```suggestion
When("the user logs in with the following credentials:", (table) => {
```

File: frontend/tests/acceptance/helpers/api/user.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+module.exports = {
+  register: async (name, email, password) => {
+    try {
+      await axios.post(`${constants.apiUrl}/auth/register`, {
+        name,
+        email,
+        password,
+      });
+    } catch (error) {
+      // throw new Error(`Cannot register user with email ${email}
```
Comment (individual-it) [inline frontend/tests/acceptance/helpers/api/user.js:13]: if its needed use the code, if not delete it, don't leave commented code in the PR

File: frontend/tests/acceptance/helpers/api/user.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    try {
+      await axios
+        .post(`${constants.apiUrl}/auth/login/`, {
+          email,
+          password,
+        })
+        .then((data) => console.log(data.data));
+    } catch (error) {
+      throw new Error(`Cannot login user with email ${email}
+     Status code: ${error.response.status}
```
Comment (individual-it) [inline frontend/tests/acceptance/helpers/api/user.js:28]: wrong indent, also in other palces

File: frontend/tests/acceptance/step_definitions/loginContext.js [hunk 6]
```diff
@@ -1,19 +1,54 @@
...
+  email = table.rows[0].cells[1].value;
+  password = table.rows[1].cells[1].value;
   loginPage.login(email, password);
 });
 
-Then('user should be redirected to the dashboard page', () => I.seeElement(dashboard.dashboardContainer));
+Then("an error message {string} should be displayed", (msg) => {
```
Comment (phil-davis) [inline frontend/tests/acceptance/step_definitions/loginContext.js:40]: Is this step being used yet?

File: frontend/tests/acceptance/pages/loginPage.js [hunk 7]
```diff
@@ -3,13 +3,33 @@ const { I } = inject();
 module.exports = {
   url: '/login',
   fields: {
-    email: '//label[contains(text(),"email")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
-    password: '//label[contains(text(),"password")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
+    email: '//label[contains(text(),"Email")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
```
Comment (individual-it) [inline frontend/tests/acceptance/pages/loginPage.js:6]:
```suggestion
    email: '//label[contains(text(),"Email")]/parent:://input[contains(@class, "MuiInputBase-input")]',
```
similar for password

File: frontend/tests/acceptance/pages/loginPage.js [hunk 8]
```diff
@@ -3,13 +3,33 @@ const { I } = inject();
...
+    password: '//label[contains(text(),"Password")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
+  },
+  buttons: {
+    login: "//button[@type='submit']",
```
Comment (individual-it) [inline frontend/tests/acceptance/pages/loginPage.js:10]: any reason that is not in `fields`?

File: frontend/tests/acceptance/pages/loginPage.js [hunk 9]
```diff
@@ -3,13 +3,33 @@ const { I } = inject();
...
   },
   login(email, password) {
+    this.fillEmail(email);
+    this.fillPassword(password);
+    this.clickLogin();
+  },
+  text: {
+    error: "//form/p[@class='makeStyles-error-5']",
```
Comment (individual-it) [inline frontend/tests/acceptance/pages/loginPage.js:18]: any reason that is not in `fields`?

File: frontend/tests/acceptance/step_definitions/loginContext.js [hunk 10]
```diff
@@ -1,19 +1,54 @@
...
+  I.seeElement(dashboard.dashboardContainer)
+);
 
-When('the user logs in with email {string} and password {string} using the webUI', (email, password) => {
+Given("the user has browsed to the login page", () => {
+  I.amOnPage(loginPage.url);
+});
+
+When("user logs in with following credentials:", (table) => {
+  email = table.rows[0].cells[1].value;
```
Comment (individual-it) [inline frontend/tests/acceptance/step_definitions/loginContext.js:35]: is it possible to use field names and not numbers?

File: frontend/tests/acceptance/step_definitions/loginContext.js (hunk 6 above)
Comment (bibekthapa619) [reply frontend/tests/acceptance/step_definitions/loginContext.js:40]: it is being used in ```validLogin``` scenario in login.feature

File: frontend/tests/acceptance/pages/loginPage.js (hunk 8 above)
Comment (bibekthapa619) [reply frontend/tests/acceptance/pages/loginPage.js:10]: tried to seperate clickable ui items from input fields

File: frontend/tests/acceptance/pages/loginPage.js (hunk 7 above)
Comment (bibekthapa619) [reply frontend/tests/acceptance/pages/loginPage.js:6]: the xpath willnot be valid
```//label[contains(text(),"Email")]/parent::div//input[contains(@class, "MuiInputBase-input")]``` can be used

File: frontend/tests/acceptance/pages/loginPage.js (hunk 9 above)
Comment (bibekthapa619) [reply frontend/tests/acceptance/pages/loginPage.js:18]: tried to seperate clickable ui items from info texts

File: frontend/tests/acceptance/features/webUISignup/signup.feature [hunk 11]
```diff
@@ -2,8 +2,9 @@ Feature: signup a user
   As a user
   I want to signup and create my account
 
+  @userSignup
   Scenario: signup a user
     Given the user has browsed to the homepage
     When the user browses to the signup page using the webUI
     And the user signs up with name "user0", email "user0@gmail.com" password "password" and confirmation password "password" using the webUI
-    Then the user should be able to login with email "user0@gmail.com" and password "password"
```
Comment (kiranparajuli589) [inline frontend/tests/acceptance/features/webUISignup/signup.feature:9]: this then is not needed to be deleted. I think its reasonable to put as it was placed!

File: frontend/tests/acceptance/helpers/api/user.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      });
+    } catch (error) {
+        throw new Error(`Cannot register user with email ${email}
+        Status code: ${error.response.status}
+        Status: ${error.response.statusText}`);
+    }
+  },
+  login: async (email, password) => {
+    try {
+      await axios
```
Comment (kiranparajuli589) [inline frontend/tests/acceptance/helpers/api/user.js:20]: await is not needed if you're are using .then() callbacks

File: frontend/tests/acceptance/pages/signupPage.js [hunk 13]
```diff
@@ -1,19 +1,22 @@
...
+  url: "/signup",
   fields: {
-    name: '//label[contains(text(),"name")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
-    email: '//label[contains(text(),"email")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
-    password: '//label[contains(text(),"password")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
-    confirmPassword: '//label[contains(text(),"confirm password")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
+    name: '//label[contains(text(),"Name")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
+    email: '//label[contains(text(),"Email")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
+    password: '//label[contains(text(),"Password")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
+    confirmPassword: '//label[contains(text(),"Confirm password")]/parent::div/div/input[contains(@class, "MuiInputBase-input")]',
```
Comment (kiranparajuli589) [inline frontend/tests/acceptance/pages/signupPage.js:9]:
```suggestion
    name: '//label[contains(text(),"Name")]/parent::div//input[contains(@class, "MuiInputBase-input")]',
    email: '//label[contains(text(),"Email")]/parent::div//input[contains(@class, "MuiInputBase-input")]',
    password: '//label[contains(text(),"Password")]/parent::div//input[contains(@class, "MuiInputBase-input")]',
    confirmPassword: '//label[contains(text(),"Confirm password")]/parent::div//input[contains(@class, "MuiInputBase-input")]',
```

File: frontend/tests/acceptance/step_definitions/loginContext.js [hunk 14]
```diff
@@ -1,19 +1,56 @@
...
+When("the user logs in with the following credentials:", (table) => {
+  let tableData = {};
+  table.parse().raw().forEach(function(data) { tableData[data[0]] = data[1] });
+  email = tableData.email;
+  password = tableData.password;
   loginPage.login(email, password);
 });
 
-Then('user should be redirected to the dashboard page', () => I.seeElement(dashboard.dashboardContainer));
+Then("an error message {string} should be displayed", (msg) => {
```
Comment (kiranparajuli589) [inline frontend/tests/acceptance/step_definitions/loginContext.js:42]:
```suggestion
Then("an error message {string} should be displayed", (message) => {
```

File: frontend/tests/acceptance/step_definitions/loginContext.js [hunk 15]
```diff
@@ -1,19 +1,56 @@
...
+  I.seeElement(loginPage.text.error);
+  I.see(msg, loginPage.text.error);
```
Comment (kiranparajuli589) [inline frontend/tests/acceptance/step_definitions/loginContext.js:44]:
```suggestion
  I.see(message, loginPage.text.error);
```

File: frontend/tests/acceptance/step_definitions/loginContext.js [hunk 16]
```diff
@@ -1,19 +1,56 @@
...
+});
+
+Then("the user should stay on the login page", () => {
+  loginPage.amOnThisPage();
+});
+
+After(() => {
+  globals.users.forEach(async (element) => {
```
Comment (kiranparajuli589) [inline frontend/tests/acceptance/step_definitions/loginContext.js:52]:
```suggestion
  globals.users.forEach(async (user) => {
```

File: frontend/tests/acceptance/features/webUISignup/signup.feature (hunk 11 above)
Comment (bibekthapa619) [reply frontend/tests/acceptance/features/webUISignup/signup.feature:9]: after signup the user is directed to dashboard so the user is auto logged in after successful signup

File: frontend/tests/acceptance/step_definitions/loginContext.js [hunk 17]
```diff
@@ -1,19 +1,56 @@
...
+);
 
-When('the user logs in with email {string} and password {string} using the webUI', (email, password) => {
+Given("the user has browsed to the login page", () => {
+  I.amOnPage(loginPage.url);
+});
+
+When("the user logs in with the following credentials:", (table) => {
+  let tableData = {};
+  table.parse().raw().forEach(function(data) { tableData[data[0]] = data[1] });
```
Comment (individual-it) [inline frontend/tests/acceptance/step_definitions/loginContext.js:36]: using hashes might make the whole thing easier https://codecept.io/bdd/#advanced-gherkin

Comment (bibekthapa619) [reply frontend/tests/acceptance/step_definitions/loginContext.js:36]: type of data table we used doesn't actually matches the desired output of hashes function.
Would it be better if data table format is changed??

Comment (individual-it) [reply frontend/tests/acceptance/step_definitions/loginContext.js:36]: If the table does not get too wide, I think I would rather make a table with a heading line and the values below

File: frontend/tests/acceptance/features/webUILogin/login.feature [hunk 18]
```diff
@@ -1,9 +1,35 @@
...
+               | email    |          |
+               | password | password |
+          Then the user should stay on the login page
+
+     @blankPasswordLogin
+     Scenario: login with blank password
+          When the user logs in with the following credentials:
+               | email    | user1@gmail.com |
+               | password |                 |
+          Then the user should stay on the login page
```
Comment (swoichha) [inline frontend/tests/acceptance/features/webUILogin/login.feature:34]: IMO we can write both these scenarios into a single scenario. You can do something like this
```suggestion
     Scenario: login with blank credentials
          When the user logs in with the following credentials:
          | email           | password           |
          | user1@gmail.com |                    |
          |                 | password           |
          |                 |                    |
          Then the user should stay on the login page

Comment (bibekthapa619) [reply frontend/tests/acceptance/features/webUILogin/login.feature:34]: Multiple user data on data table won't probably solve it. ```Then``` statement will execute only once even if there are multiple users.

File: frontend/tests/acceptance/features/webUILogin/login.feature [hunk 19]
```diff
@@ -1,9 +1,35 @@
...
+          And the user has browsed to the login page
+
+     @validLogin
+     Scenario: login a valid user
+          When the user logs in with email "user1@gmail.com" and password "password" using the webUI
+          Then the user should be redirected to the dashboard page
+
+     @invalidLogin
+     Scenario: login with invalid credentials
+          When the user logs in with the following credentials:
```
Comment (HariBhandari07) [inline frontend/tests/acceptance/features/webUILogin/login.feature:17]: For webUI tests, append `using the webUI` at the end

File: frontend/tests/acceptance/features/webUILogin/login.feature [hunk 20]
```diff
@@ -1,9 +1,35 @@
...
+               | email           | password |
+               | user1@gmail.com | passwo   |
+          Then the user should stay on the login page
+
+     @blankEmailLogin
+     Scenario: login with blank email
+          When the user logs in with the following credentials:
```
Comment (HariBhandari07) [inline frontend/tests/acceptance/features/webUILogin/login.feature:24]: same as above

File: frontend/tests/acceptance/features/webUILogin/login.feature [hunk 21]
```diff
@@ -1,9 +1,35 @@
...
-    Given the user has signed up with name "user1", email "user1@gmail.com" password "password"
-    When the user browses to the login page
-    And the user logs in with email "user1@gmail.com" and password "password" using the webUI
-    Then user should be redirected to the dashboard page
+     As a user
+     I want to login to my dashboard
+     So that I can access jobtriage services
+
+     Background:
+          Given the user has signed up with name "user1", email "user1@gmail.com" password "password"
```
Comment (Talank) [inline frontend/tests/acceptance/features/webUILogin/login.feature:7]:
```suggestion
          Given the user has signed up with name "user1", email "user1@gmail.com" and password "password"
```

**Comment by kiranparajuli589** [comment]
LGTM :+1:

**Pull request closed without merging**